src/openfermion/ops/_code_operator.py

```diff
@@ -0,0 +1,298 @@
+import copy
+
+import numpy
+import scipy
+import scipy.sparse
+
+from openfermion.ops import SymbolicBinary
+
+
+def _shift_decoder(decode, shift_constant):
+    """ Shifts the indices of a decoder by a constant.
+    
+    Args:
+        decode (list of SymbolicBinary): a decoder
+        shift_constant (int): the qubit index that corresponds to the offset.
+
+    Returns (list of SymbolicBinary):  shifted decoder
+    """
+    decode_shifted = []
+    for entry in decode:
+        tmp_entry = copy.deepcopy(entry)
+        tmp_entry._shift(shift_constant)
+        decode_shifted.append(tmp_entry)
+    return numpy.array(decode_shifted)
+
+
+def _double_decoding(dec1, dec2):
+    """ Concatenates two decodings
+    
+    Args:
+        dec1 (list of SymbolicBinary): decoding of the outer code layer
+        dec2 (list of SymbolicBinary): decoding of the inner code layer
+    
+    Returns (list of SymbolicBinary): the decoding defined by
+    w -> dec1(dec2(w))
+    """
+    doubled_decoder = []
+    for entry in dec1:
+        tmp_sum = 0
+        for summand in entry.terms:
+            tmp_term = SymbolicBinary('1')
+            for factor in summand:
+                if factor[1] == 'W':
+                    tmp_term *= dec2[factor[0]]
+            tmp_sum = tmp_term + tmp_sum
+        doubled_decoder += [tmp_sum]
+    return numpy.array(doubled_decoder)
+
+
+def linearize_decoder(mtx):
+    """ Outputs row_idx linear decoding function from row_idx matrix
+
+    Args:
+        mtx (array or list): row_idx 2D list of lists or row_idx numpy array
+         to derive the decoding function from
+
+    Returns: list of SymbolicBinary
+    """
+    mtx = numpy.array(list(map(numpy.array, mtx)))
+    system_dim, code_dim = numpy.shape(mtx)
+    res = [] * system_dim
+    for row_idx in numpy.arange(system_dim):
+        dec_str = ''
+        for col_idx in numpy.arange(code_dim):
+            if mtx[row_idx, col_idx] == 1:
+                dec_str += 'W' + str(col_idx) + ' + '
+        dec_str = dec_str.rstrip(' + ')
+        res.append(SymbolicBinary(dec_str))
+    return numpy.array(res)
+
+
+class BinaryCodeError(Exception):
+    pass
+
+
+class BinaryCode(object):
```

Comment:
This is fairly comprehensive docs. Even though I think I forgot to request this on your SymbolicBinary class, we usually have "attributes" listed with descriptions in the doc strings under the class definition. See (https://github.com/quantumlib/OpenFermion/blob/master/src/openfermion/ops/_symbolic_operator.py) for an example.